keps/sig-scheduling/3633-matchlabelkeys-to-podaffinity/README.md

```diff
@@ -0,0 +1,929 @@
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `MatchLabelKeysInPodAffinityAndPodAntiAffinity`
+  - Components depending on the feature gate: `kube-scheduler`, `kube-apiserver`
+- [x] Other
+  - Describe the mechanism: when `PodAffinityTerm` has `MatchLabelKeys`, The scheduler will use those keys to look up label values from the incoming pod; and those key-value labels are ANDed with `LabelSelector` (the existing field). 
+  - Will enabling / disabling the feature require downtime of the control
+    plane?: No.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).: No.
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+The feature can be disabled in Alpha and Beta versions
+by restarting kube-apiserver and kube-scheduler with the feature-gate off.
+In terms of Stable versions, users can choose to opt-out by not setting the
+`MatchLabelKeys` field.
+
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Scheduling of new Pods is affected. (Scheduled Pods aren't affected.)
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+No - Only unit and integration tests will be added.
```

Comment:
For features that introduce new API fields, we generally recomment to introduce an API-level test - see the comment above:
```
Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
```
Can you please add such test as part of Alpha (and for now just switch it here to mention that such test will be added)?

Reply:
Sure.